File: android/app/src/main/java/com/appxstudios/festivalconnection/mesh/nostr/NostrEvent.kt
package com.appxstudios.festivalconnection.mesh.nostr

import com.appxstudios.festivalconnection.security.NostrIdentity
import com.google.gson.Gson
import com.google.gson.annotations.SerializedName
import java.security.MessageDigest

data class NostrEvent(
    val id: String,
    val pubkey: String,
    @SerializedName("created_at") val createdAt: Long,
    val kind: Int,
    val tags: List<List<String>>,
    val content: String,
    val sig: String
) {
    companion object {
        private val gson = Gson()

        /**
         * NIP-01 canonical serialization: [0, pubkey, created_at, kind, tags, content].
         * Escapes per JSON spec: ", \, \b, \f, \n, \r, \t, and all control chars (< 0x20)
         * as \u00XX. Forward slash is NOT escaped. This matches iOS JSONSerialization with
         * .withoutEscapingSlashes, ensuring identical event IDs across platforms.
         */
        fun serialize(pubkey: String, createdAt: Long, kind: Int, tags: List<List<String>>, content: String): String {
            val tagsJson = tags.joinToString(",") { tag ->
                "[" + tag.joinToString(",") { "\"${escapeJsonString(it)}\"" } + "]"
            }
            return "[0,\"${escapeJsonString(pubkey)}\",$createdAt,$kind,[$tagsJson],\"${escapeJsonString(content)}\"]"
        }

        fun computeId(pubkey: String, createdAt: Long, kind: Int, tags: List<List<String>>, content: String): String {
            val serialized = serialize(pubkey, createdAt, kind, tags, content)
            val digest = MessageDigest.getInstance("SHA-256")
            val hash = digest.digest(serialized.toByteArray(Charsets.UTF_8))
            return NostrIdentity.bytesToHex(hash)
        }

        fun create(kind: Int, content: String, tags: List<List<String>> = emptyList()): NostrEvent {
            val pubkey = NostrIdentity.publicKeyHex
            val createdAt = System.currentTimeMillis() / 1000
            val id = computeId(pubkey, createdAt, kind, tags, content)
            val idBytes = NostrIdentity.hexToBytes(id) ?: ByteArray(32)
            val sigBytes = NostrIdentity.sign(idBytes)
            val sig = NostrIdentity.bytesToHex(sigBytes)

            return NostrEvent(id, pubkey, createdAt, kind, tags, content, sig)
        }

        fun fromJson(json: String): NostrEvent? {
            return try { gson.fromJson(json, NostrEvent::class.java) } catch (e: Exception) { null }
        }

        /**
         * NIP-01 / RFC 8259 compliant string escaping for canonical JSON.
         * Escapes: " \ \b \f \n \r \t and all control characters (code < 0x20) as \u00XX.
         * Does NOT escape forward slash (/) — slashes pass through unchanged.
         */
        private fun escapeJsonString(s: String): String {
            val sb = StringBuilder(s.length + 8)
            for (c in s) {
                when (c) {
                    '"' -> sb.append("\\\"")
                    '\\' -> sb.append("\\\\")
                    '\b' -> sb.append("\\b")
                    '' -> sb.append("\\f")
                    '\n' -> sb.append("\\n")
                    '\r' -> sb.append("\\r")
                    '\t' -> sb.append("\\t")
                    else -> {
                        if (c.code < 0x20) {
                            sb.append(String.format("\\u%04x", c.code))
                        } else {
                            sb.append(c)
                        }
                    }
                }
            }
            return sb.toString()
        }
    }

    fun verifyId(): Boolean {
        val computed = computeId(pubkey, createdAt, kind, tags, content)
        return computed == id
    }

    /**
     * Verify the BIP-340 Schnorr signature against the event ID and pubkey.
     * Mirrors iOS NostrEvent.verifySignature (using P256K.Schnorr.XonlyKey.isValid).
     * Returns false if signature is missing, malformed, or does not validate.
     */
    fun verifySignature(): Boolean {
        return try {
            val pubBytes = NostrIdentity.hexToBytes(pubkey) ?: return false
            if (pubBytes.size != 32) return false
            val sigBytes = NostrIdentity.hexToBytes(sig) ?: return false
            if (sigBytes.size != 64) return false
            val msgHash = NostrIdentity.hexToBytes(id) ?: return false
            if (msgHash.size != 32) return false
            NostrIdentity.verifyBIP340(pubBytes, msgHash, sigBytes)
        } catch (e: Exception) {
            false
        }
    }

    fun toJson(): String = gson.toJson(this)
}

File: ios/FestivalConnection/Mesh/Nostr/NostrEvent.swift
import Foundation
import CryptoKit
import P256K

struct NostrEvent: Codable, Identifiable, Equatable {
    let id: String
    let pubkey: String
    let createdAt: Int64
    let kind: Int
    let tags: [[String]]
    let content: String
    let sig: String

    enum CodingKeys: String, CodingKey {
        case id, pubkey
        case createdAt = "created_at"
        case kind, tags, content, sig
    }

    static func == (lhs: NostrEvent, rhs: NostrEvent) -> Bool {
        lhs.id == rhs.id
    }

    /// Serialize for ID computation per NIP-01: [0, pubkey, created_at, kind, tags, content]
    /// Uses JSONSerialization with withoutEscapingSlashes for spec-compliant encoding,
    /// including \b, \f, \u00XX control characters, and proper UTF-8 handling.
    static func serializedData(pubkey: String, createdAt: Int64, kind: Int, tags: [[String]], content: String) -> Data {
        let serialized: [Any] = [
            0,
            pubkey,
            createdAt,
            kind,
            tags,
            content
        ]
        guard let data = try? JSONSerialization.data(
            withJSONObject: serialized,
            options: [.withoutEscapingSlashes]
        ) else {
            return Data()
        }
        return data
    }

    /// Compute event ID as SHA-256 hex of the canonically serialized event.
    static func computeId(pubkey: String, createdAt: Int64, kind: Int, tags: [[String]], content: String) -> String {
        let data = serializedData(pubkey: pubkey, createdAt: createdAt, kind: kind, tags: tags, content: content)
        let hash = SHA256.hash(data: data)
        return hash.compactMap { String(format: "%02x", $0) }.joined()
    }

    /// Create and sign a new event
    @MainActor
    static func create(kind: Int, content: String, tags: [[String]] = []) -> NostrEvent {
        let pubkey = NostrIdentity.shared.publicKeyHex
        let createdAt = Int64(Date().timeIntervalSince1970)
        let id = computeId(pubkey: pubkey, createdAt: createdAt, kind: kind, tags: tags, content: content)
        let idData = hexToData(id) ?? Data()
        let sigData = NostrIdentity.shared.sign(idData)
        let sig = sigData.map { String(format: "%02x", $0) }.joined()

        return NostrEvent(
            id: id,
            pubkey: pubkey,
            createdAt: createdAt,
            kind: kind,
            tags: tags,
            content: content,
            sig: sig
        )
    }

    /// Verify event ID matches content
    func verifyId() -> Bool {
        let computed = NostrEvent.computeId(pubkey: pubkey, createdAt: createdAt, kind: kind, tags: tags, content: content)
        return computed == id
    }

    /// Verify the BIP-340 Schnorr signature against the event ID and pubkey.
    /// Returns false if signature is missing, malformed, or does not validate.
    func verifySignature() -> Bool {
        guard let sigData = NostrEvent.hexToData(sig),
              sigData.count == 64,
              let pubData = NostrEvent.hexToData(pubkey),
              pubData.count == 32,
              let idData = NostrEvent.hexToData(id),
              idData.count == 32 else {
            return false
        }

        // The Nostr event ID IS the SHA-256 message hash that was Schnorr-signed.
        guard let signature = try? P256K.Schnorr.SchnorrSignature(dataRepresentation: sigData) else {
            return false
        }

        var messageBytes = [UInt8](idData)
        let xonly = P256K.Schnorr.XonlyKey(dataRepresentation: pubData)
        return xonly.isValid(signature, for: &messageBytes)
    }

    /// Convert to JSON for sending
    func toJSON() -> String {
        let encoder = JSONEncoder()
        encoder.outputFormatting = []
        guard let data = try? encoder.encode(self) else { return "{}" }
        return String(data: data, encoding: .utf8) ?? "{}"
    }

    private static func hexToData(_ hex: String) -> Data? {
        var data = Data()
        var temp = hex
        while temp.count >= 2 {
            let s = String(temp.prefix(2))
            temp = String(temp.dropFirst(2))
            guard let b = UInt8(s, radix: 16) else { return nil }
            data.append(b)
        }
        return data
    }
}
